### PR TITLE
Pchain - Cleanup NodeID generation in UTs

### PR DESCRIPTION
**Pull request by abi87**
## Why this should be merged
In platformVM UTs, we have the habit of generating nodeIDs from addresses with something like `nodeID := ids.NodeID(keys[0].PublicKey().Address())`. In the vast majority of cases, this is not really needed and using `ids.GenerateTestNodeID()` is cleaner.
Moreover the first expression cast the address type (ids.ShortID) to an ids.NodeID, which won't be possible anymore upon acceptance of  [ed25519 keys](https://github.com/avalanche-foundation/ACPs/blob/main/ACPs/20-ed25519-p2p.md).
So this PR just clean some stuff up

## How this works
Mostly replaces:
```
nodeID := ids.NodeID(key.PublicKey().Address()) --> nodeID := ids.GenerateTestNodeID
```

In the few cases where we need to refer to specific nodeIDs (e.g. those of genesis validators) we do it introducing genesisNodeID variable

## How this was tested
CI

### CODE REVIEW

File: vms/platformvm/block/executor/proposal_block_test.go [hunk 1]
```diff
@@ -392,57 +392,46 @@ func TestBanffProposalBlockUpdateStakers(t *testing.T) {
 	// Staker5:                                     |--------------------|
 
 	// Staker0 it's here just to allow to issue a proposal block with the chosen endTime.
-	staker0RewardAddress := ids.ShortID{2}
 	staker0 := staker{
-		nodeID:        ids.NodeID(staker0RewardAddress),
-		rewardAddress: staker0RewardAddress,
+		nodeID:        ids.BuildTestNodeID([]byte{0xf0}),
```
Comment (abi87) [inline vms/platformvm/block/executor/proposal_block_test.go:396]: This is a nasty test where multiple stakers may join and leave the staker set. The order in which they do it is asserted (so wrong order will make test fail) but it may depend on the staker.TxID, which in turns depend on *every* feature of the transaction creating the staker.
Here I avoid using ids.GenerateTestNodeID because it wil give different result depending on the order in which UTs are executed. Hard-coding nodeID as I did here should make the test more robust

File: vms/platformvm/block/executor/standard_block_test.go [hunk 2]
```diff
@@ -364,38 +364,38 @@ func TestBanffStandardBlockUpdateStakers(t *testing.T) {
 	// Staker4:            |------------------------|
 	// Staker5:                                 |--------------------|
 	staker1 := staker{
-		nodeID:        ids.GenerateTestNodeID(),
-		rewardAddress: ids.GenerateTestShortID(),
+		nodeID:        ids.BuildTestNodeID([]byte{0xf1}),
```
Comment (abi87) [inline vms/platformvm/block/executor/standard_block_test.go:367]: Aligned this test with proposal block one above. After all they test the very same things, just on different block types

File: vms/platformvm/block/executor/helpers_test.go [hunk 3]
```diff
@@ -81,9 +81,19 @@ var (
...
 	errMissing = errors.New("missing")
 )
 
+func init() {
+	for range preFundedKeys {
+		nodeID := ids.GenerateTestNodeID()
+		genesisNodeIDs = append(genesisNodeIDs, nodeID)
+	}
+}
+
```
Comment (dhrubabasu) [inline vms/platformvm/block/executor/helpers_test.go:96]: nit: we should pre-allocate this

```suggestion
func init() {
	genesisNodeIDs = make([]ids.NodeID, len(preFundedKeys))
	for i, _ := range preFundedKeys {
		genesisNodeIDs[i] = ids.GenerateTestNodeID()
	}
}
```

File: vms/platformvm/block/executor/proposal_block_test.go [hunk 4]
```diff
@@ -546,10 +535,7 @@ func TestBanffProposalBlockUpdateStakers(t *testing.T) {
 			advanceTimeTo: []time.Time{staker1.startTime, staker2.startTime, staker3.startTime, staker5.startTime},
 			expectedStakers: map[ids.NodeID]stakerStatus{
 				staker1.nodeID: current,
-
-				// given its txID, staker2 will be
-				// rewarded and moved out of current stakers set
-				// staker2.nodeID: current,
+				staker2.nodeID: current,
```
Comment (dhrubabasu) [inline vms/platformvm/block/executor/proposal_block_test.go:538]: Why is this being changed?

File: vms/platformvm/txs/validator_test.go [hunk 5]
```diff
@@ -9,22 +9,21 @@ import (
...
-
 func TestBoundedBy(t *testing.T) {
 	require := require.New(t)
 
+	nodeID := ids.GenerateTestNodeID()
+
 	// case 1: a starts, a finishes, b starts, b finishes
 	aStartTime := uint64(0)
 	aEndTIme := uint64(1)
+
```
Comment (dhrubabasu) [inline vms/platformvm/txs/validator_test.go:24]: nit: remove this new line

File: vms/platformvm/block/builder/helpers_test.go [hunk 6]
```diff
@@ -76,9 +76,19 @@ var (
...
+
 	errMissing = errors.New("missing")
 )
 
+func init() {
+	for range preFundedKeys {
+		nodeID := ids.GenerateTestNodeID()
+		genesisNodeIDs = append(genesisNodeIDs, nodeID)
+	}
+}
```
Comment (dhrubabasu) [inline vms/platformvm/block/builder/helpers_test.go:90]: nit: we should pre-allocate this

```suggestion
func init() {
	genesisNodeIDs = make([]ids.NodeID, len(preFundedKeys))
	for i, _ := range preFundedKeys {
		genesisNodeIDs[i] = ids.GenerateTestNodeID()
	}
}
```

File: vms/platformvm/txs/executor/helpers_test.go [hunk 7]
```diff
@@ -73,9 +73,19 @@ var (
...
+
 	errMissing = errors.New("missing")
 )
 
+func init() {
+	for range preFundedKeys {
+		nodeID := ids.GenerateTestNodeID()
+		genesisNodeIDs = append(genesisNodeIDs, nodeID)
+	}
+}
```
Comment (dhrubabasu) [inline vms/platformvm/txs/executor/helpers_test.go:87]: nit: we should pre-allocate this

```suggestion
func init() {
	genesisNodeIDs = make([]ids.NodeID, len(preFundedKeys))
	for i, _ := range preFundedKeys {
		genesisNodeIDs[i] = ids.GenerateTestNodeID()
	}
}
```

File: ids/test_generator.go [hunk 8]
```diff
@@ -19,6 +19,15 @@ func GenerateTestShortID() ShortID {
 	return newShortID
 }
 
+// BuildTestNodeID is an utility to build NodeID from bytes in UTs
+// It must not be used in production code. In production code we should
+// use ToNodeID, which performs proper length checking.
```
[the remark this answers is not in the document]
Comment (abi87) [reply ids/test_generator.go]: Done!

File: vms/platformvm/block/executor/helpers_test.go (hunk 3 above)
Comment (abi87) [reply vms/platformvm/block/executor/helpers_test.go]: yup, done

File: vms/platformvm/txs/validator_test.go (hunk 5 above)
Comment (abi87) [reply vms/platformvm/txs/validator_test.go]: Done

File: vms/platformvm/block/builder/helpers_test.go (hunk 6 above)
Comment (abi87) [reply vms/platformvm/block/builder/helpers_test.go]: Done

File: vms/platformvm/txs/executor/helpers_test.go (hunk 7 above)
Comment (abi87) [reply vms/platformvm/txs/executor/helpers_test.go]: Done

File: vms/platformvm/block/executor/proposal_block_test.go (hunk 4 above)
Comment (abi87) [reply vms/platformvm/block/executor/proposal_block_test.go]: Hey @dhrubabasu thanks for spotting this.
This is another way this test is brittle: it asserts that stakers are there if the right chain time arrives, but it does not work with stakers removal. Specifically in this test case, we reach the chain-time where staker5 should be added (which happens and it is duly checked among test assertion) *and* staker2 should be remove, which does not happen.
In the interest of scoping this PR, I carried out the minimal number of changes. Will make another PR to fix/improve this test!

File: vms/platformvm/block/executor/standard_block_test.go [hunk 9]
```diff
@@ -474,12 +474,11 @@ func TestBanffStandardBlockUpdateStakers(t *testing.T) {
 			},
 		},
 		{
-			description:   "advance time to staker5 end",
+			description:   "advance time to staker5 start",
 			stakers:       []staker{staker1, staker2, staker3, staker4, staker5},
 			advanceTimeTo: []time.Time{staker1.startTime, staker2.startTime, staker3.startTime, staker5.startTime},
 			expectedStakers: map[ids.NodeID]stakerStatus{
 				staker1.nodeID: current,
-				staker2.nodeID: current,
```
Comment (abi87) [inline vms/platformvm/block/executor/standard_block_test.go:482]: @dhrubabasu another instance of the same brittleness we have in `TestBanffProposalBlockUpdateStakers` test. Staker2 should be evicted here, so not to appear among expected stakers. It is not evicted since the test does not generate the proposalBlockto do that. For the time being i just drop staker2 from the expected stakers. Will make another PR to fix this test.

File: vms/platformvm/block/executor/proposal_block_test.go [hunk 10]
```diff
@@ -541,15 +530,11 @@ func TestBanffProposalBlockUpdateStakers(t *testing.T) {
...
-			description:   "advance time to staker5 end",
+			description:   "advance time to staker5 start",
 			stakers:       []staker{staker1, staker2, staker3, staker4, staker5},
 			advanceTimeTo: []time.Time{staker1.startTime, staker2.startTime, staker3.startTime, staker5.startTime},
 			expectedStakers: map[ids.NodeID]stakerStatus{
 				staker1.nodeID: current,
-
-				// given its txID, staker2 will be
-				// rewarded and moved out of current stakers set
-				// staker2.nodeID: current,
```
Comment (abi87) [inline vms/platformvm/block/executor/proposal_block_test.go:552]: the comment is not really true. We don't generate a proposal block with RewardValidatorTx to remove staker2

File: vms/platformvm/block/executor/proposal_block_test.go (hunk 1 above)
Comment (danlaine) [reply vms/platformvm/block/executor/proposal_block_test.go:396]: ^ Probably worth including this in a code comment

**Comment by danlaine** [comment]
I think changing the node ID generation to not be dependent on keys makes sense but I left a few nits

File: vms/platformvm/vm_test.go [hunk 11]
```diff
@@ -119,6 +122,15 @@ var (
 	errMissing = errors.New("missing")
 )
 
+func init() {
+	for _, key := range keys {
+		nodeBytes := key.PublicKey().Address()
```
[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/vm_test.go:127]: Because of the `TestGetStake` test in `vms/platformvm/service_test.go` which does some checks against genesis content. I think we should eventually clean it up, just chose not to do it in this PR

[the remark this answers is not in the document]
Comment (danlaine) [reply vms/platformvm/vm_test.go:127]: Could we put a TODO?

[the remark this answers is not in the document]
Comment (abi87) [reply vms/platformvm/vm_test.go:129]: Done

**Review by danlaine — APPROVED** [review]

**Review by dhrubabasu — APPROVED** [review]

File: ids/test_generator.go [hunk 12]
```diff
@@ -19,6 +19,15 @@ func GenerateTestShortID() ShortID {
...
 
+// BuildTestShortID is an utility to build ShortID from bytes in UTs
+// It must not be used in production code. In production code we should
+// use ToShortID, which performs proper length checking.
+func BuildTestShortID(src []byte) ShortID {
+	res := ShortID{}
+	copy(res[:], src)
+	return res
+}
+
```
Comment (StephenButtolph) [inline ids/test_generator.go:30]: Why are we adding this function? Can't we just directly specify the `ShortID`?

File: vms/platformvm/block/executor/proposal_block_test.go [hunk 13]
```diff
@@ -392,57 +392,52 @@ func TestBanffProposalBlockUpdateStakers(t *testing.T) {
...
+	// In this test multiple stakers may join and leave the staker set at the same time.
+	// The order in which they do it is asserted; the order may depend on the staker.TxID,
+	// which in turns depend on every feature of the transaction creating the staker.
+	// So in this test we avoid ids.GenerateTestNodeID, in favour of ids.BuildTestNodeID
+	// so that TxID does not depend on the order we run tests.
 	staker0 := staker{
-		nodeID:        ids.NodeID(staker0RewardAddress),
-		rewardAddress: staker0RewardAddress,
+		nodeID:        ids.BuildTestNodeID([]byte{0xf0}),
+		rewardAddress: ids.BuildTestShortID([]byte{0xf0}),
```
Comment (StephenButtolph) [inline vms/platformvm/block/executor/proposal_block_test.go:403]:
```suggestion
		rewardAddress: ids.ShortID{0xf0},
```
Seems simpler no? (same below)

File: vms/platformvm/vm_regression_test.go [hunk 14]
```diff
@@ -1111,12 +1103,12 @@ func TestValidatorSetAtCacheOverwriteRegression(t *testing.T) {
...
-		nodeID2: defaultWeight,
-		nodeID3: defaultWeight,
-		nodeID4: defaultWeight,
-		nodeID5: vm.MaxValidatorStake,
+		nodeIDs[0]: defaultWeight,
+		nodeIDs[1]: defaultWeight,
+		nodeIDs[2]: defaultWeight,
+		nodeIDs[3]: defaultWeight,
+		nodeIDs[4]: defaultWeight,
+		nodeID5:    vm.MaxValidatorStake,
```
Comment (StephenButtolph) [inline vms/platformvm/vm_regression_test.go:1111]: do we still want to call it `nodeID5` if we don't have the other `nodeID`s?

File: vms/platformvm/vm_regression_test.go [hunk 15]
```diff
@@ -1002,22 +999,17 @@ func TestValidatorSetAtCacheOverwriteRegression(t *testing.T) {
 		vm.ctx.Lock.Unlock()
 	}()
 
-	nodeID0 := ids.NodeID(keys[0].PublicKey().Address())
-	nodeID1 := ids.NodeID(keys[1].PublicKey().Address())
-	nodeID2 := ids.NodeID(keys[2].PublicKey().Address())
-	nodeID3 := ids.NodeID(keys[3].PublicKey().Address())
-	nodeID4 := ids.NodeID(keys[4].PublicKey().Address())
-
+	nodeIDs := genesisNodeIDs
```
Comment (StephenButtolph) [inline vms/platformvm/vm_regression_test.go:1002]: I feel like we can just directly use `genesisNodeIDs`

File: vms/platformvm/block/executor/proposal_block_test.go [hunk 16]
```diff
@@ -541,15 +536,11 @@ func TestBanffProposalBlockUpdateStakers(t *testing.T) {
...
-			description:   "advance time to staker5 end",
+			description:   "advance time to staker5 start",
 			stakers:       []staker{staker1, staker2, staker3, staker4, staker5},
 			advanceTimeTo: []time.Time{staker1.startTime, staker2.startTime, staker3.startTime, staker5.startTime},
 			expectedStakers: map[ids.NodeID]stakerStatus{
 				staker1.nodeID: current,
-
-				// given its txID, staker2 will be
-				// rewarded and moved out of current stakers set
-				// staker2.nodeID: current,
```
Comment (StephenButtolph) [inline vms/platformvm/block/executor/proposal_block_test.go:552]: We should be marking `staker2.nodeID: current,` here

File: vms/platformvm/block/executor/standard_block_test.go [hunk 17]
```diff
@@ -474,12 +480,11 @@ func TestBanffStandardBlockUpdateStakers(t *testing.T) {
...
```
Comment (StephenButtolph) [inline vms/platformvm/block/executor/standard_block_test.go:482]: Why is this changed in this PR? asserting `staker2.nodeID: current,` is correct based on how time is advanced in this test

File: vms/platformvm/block/executor/proposal_block_test.go (hunk 13 above)
Comment (abi87) [reply vms/platformvm/block/executor/proposal_block_test.go]: It is. Dropped the helper

File: ids/test_generator.go (hunk 12 above)
Comment (abi87) [reply ids/test_generator.go]: Dropped

File: vms/platformvm/vm_regression_test.go (hunk 14 above)
Comment (abi87) [reply vms/platformvm/vm_regression_test.go]: renamed to extraNode

File: vms/platformvm/vm_regression_test.go (hunk 15 above)
Comment (abi87) [reply vms/platformvm/vm_regression_test.go]: Done

File: vms/platformvm/block/executor/standard_block_test.go (hunk 17 above)
Comment (abi87) [reply vms/platformvm/block/executor/standard_block_test.go:482]: I felt it was weird.
We should not expect staker2 to be available when chainTime reaches its end time. In prod a ProposalBlock would be eventually issued to remove it from the validator set.
In the test we'll find it since we don't issue the proposalBlock and we just assert the advance time behaviour.
I reintroduced the staker with a comment.

File: vms/platformvm/block/executor/proposal_block_test.go (hunk 16 above)
Comment (abi87) [reply vms/platformvm/block/executor/proposal_block_test.go]: Done

File: vms/platformvm/block/executor/standard_block_test.go (hunk 17 above)
Comment (StephenButtolph) [reply vms/platformvm/block/executor/standard_block_test.go:482]: I don't think it is required for a proposal block to have been issued in this case. It's valid to issue a `StandardBlock` to advance the time up to the time to remove a validator (but not actually remove them).

**Review by StephenButtolph — APPROVED** [review]
